### PR TITLE
Opal compatibility + Oga support

### PR DESCRIPTION
**Pull request by hmdne**
Opal is a Ruby-to-JavaScript transpiler that involves certain semantics that the compiled code has to follow. Most importantly, those are as follows:

* No C library extensions
* Assumption that strings are immutable
* Correct knowledge about required files before compilation takes place

Fixing those issues in code will allow us to compile Plurimath to JavaScript, so that the code will be usable directly in web browsers and environments like NodeJS. This is the work that we are doing in the `plurimath-js` repository and therefore any issues related to the JavaScript version of the library should be directed there.

One of the most important parts is that the Ox library, as of now, requires a C extension. An approach we took was to port another library that does mostly the same, namely Oga and then to port Plurimath to support both Ox and Oga. We also chose Oga, because of apparent support of Oga, along with Nokogiri, in the `equivalent-xml` library, which we use for comparing XML trees.

Support for both has been implemented by doing a façade pattern named `Plurimath.xml_engine` that uses Ox directly if needed and Oga via an API that is mostly compatible with Ox (as per usage patterns in Plurimath codebase).

The Ruby runtime still uses Ox, except when an environment variable `PLURIMATH_OGA` is set, in which case, we use Oga. We use this implementation internally to test for Oga shim completeness while still running code on MRI. If problems arise, this means that it's not Opal fault directly, that problems happen, but Oga shim issues.

Other changes most likely refer to Opal build system, minor compatibility fixes (like assumptions that strings are mutable), corrections of paths. All of those changes should keep the logic mostly the same when running in MRI.

This is not a full solution for Opal compatibility yet. There are still important parts of that work in progress. It's not expected at this point that either all Opal or Oga tests will pass. It is expected though that all the tests that are currently being ran will pass.

Ref: plurimath-js/plurimath-js#3

### CODE REVIEW

**Review by ronaldtse — APPROVED** [review]

**Comment by ronaldtse** [comment]
Thank you @hmdne , let's have @suleman-uzair and @HassanAkbar review this before merging.

Please also help fix the tests before this is mergable, thanks.

**Comment by hmdne** [comment]
> Please also help fix the tests before this is mergable, thanks.

The tests are passing, it looks like a CI failure.

**Comment by ronaldtse** [comment]
@suleman-uzair could you please help fix the "tests-passed" test since it doesn't seem to be related to the changes here? Thanks.

**Comment by suleman-uzair** [comment]
@ronaldtse I will look into it as soon as possible.
@hmdne, could you please add tests for the new code added to resolve the code coverage error?

File: lib/plurimath/math.rb [hunk 1]
```diff
@@ -19,9 +19,8 @@
 require_relative "html/parser"
 require_relative "omml/parser"
 require_relative "utility"
-require "ox"
+
```
Comment (suleman-uzair) [inline lib/plurimath/math.rb:22]: This seems to be an extra and unnecessary space.

File: lib/plurimath/xml_engine/oga.rb [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+        attr_reader :out
+
+        private
+
+        def dump_attrs(node)
+          node.attributes.map do |i|
+            #if i.namespace
+            #  %{ #{i.namespace.name}:#{i.name}="#{attr_entities i.value}"}
+            #else
```
Comment (suleman-uzair) [inline lib/plurimath/xml_engine/oga.rb:178]: Unused code should be removed.

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
+  SimpleCov.start do
+    add_filter "/spec/"
+  end
 
-require "bundler/setup"
+  require "bundler/setup"
+end
 require "plurimath"
 require "rspec/matchers"
+# require 'nokogiri'
```
Comment (suleman-uzair) [inline spec/spec_helper.rb:11]: let's remove this line if we don't need it.

File: lib/plurimath/xml_engine/oga.rb [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+              "&lt;"
+            elsif i == ORD_GT
+              "&gt;"
+            elsif i == ORD_APOS && attr
+              "&apos;"
+            elsif i == ORD_QUOT && attr
+              "&quot;"
+            elsif i == ORD_NEWLINE || i == ORD_CARRIAGERETURN
+              i.chr("utf-8")
+            elsif i == ORD_AMP || i == ORD_LT || i == ORD_GT ||
```
Comment (suleman-uzair) [inline lib/plurimath/xml_engine/oga.rb:206]: shouldn't case be used instead of if?

**Comment by hmdne** [comment]
Thank you for the review, @suleman-uzair. I will come back to this tomorrow.

As of now, I have one question and perhaps you will be one who knows the answer to that:

I am dealing with this testcase: `rspec ./spec/plurimath/omml_spec.rb:60 # Plurimath::Omml.to_latex, to_mathml, to_asciimath contains example #01 returns parsed Asciimath to Formula` - in particular with MathML output.

The diff I receive, where `-` is correct, `+` is what my (Opal) port outputs.

```diff
        <math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
          <mstyle displaystyle="true">
            <msubsup>
       -      <mo>&#x222b;</mo>
       +      <mo>int</mo>
              <mn>2</mn>
              <mn>1</mn>
            </msubsup>
```

I have tried to investigate this problem thoroughly. What I found is this. The difference stems from this particular function:

https://github.com/plurimath/plurimath/blob/main/lib/plurimath/math/symbol.rb#L28-L40

In particular, this fragment:

```ruby
          mo_value = (unicodes[value] || unicode || value).to_s
```

In the case of Opal, `unicodes[value]` is truthy, so the final result is `int`. In case of Ruby, this is falsy, so the final result is `&#x222b;`.

 value is `"&#x222b;"`. Let's find this Hash.

https://github.com/plurimath/plurimath/blob/main/lib/plurimath/mathml/constants.rb#L96

Ok - so it's quite simple why this is falsy in Ruby. `"&#x222b;": "int",` is equivalent to `        :"&#x222b;" => "int",`. Symbols are not Strings, there's only Symbol available, there's no String with such a name in this hash, so we get to return `value` which is `"&#x222b;"` and test passes.

In Opal, Strings are the same type as Symbols. So, incorrectly, this is truthy and the value of the entire expression is `"int"`, so the test fails.

I have re-ran the test suite using Ruby (not Opal) and it seems like, `value` is always a `String`, never a `Symbol` (for all test cases at least). Therefore if I just patch the file `lib/plurimath/math/symbol.rb` this way...:

```diff
diff --git a/lib/plurimath/math/symbol.rb b/lib/plurimath/math/symbol.rb
index 70f8b37..7f81670 100644
--- a/lib/plurimath/math/symbol.rb
+++ b/lib/plurimath/math/symbol.rb
@@ -32,7 +32,7 @@ module Plurimath
         unicodes = Mathml::Constants::UNICODE_SYMBOLS
         unicode = unicodes.invert[value]
         if operator?(unicode) || unicode || explicit_checks(unicode)
-          mo_value = (unicodes[value] || unicode || value).to_s
+          mo_value = (unicode || value).to_s
           return Utility.ox_element("mo") << mo_value
         end
 ```

...the Ruby-side stays equivalent and at the same time Opal passes the test.

I have a suspicion though, that this isn't what was intended. In fact, I suppose that the logic in this function may be incorrect, and what was supposed to happen was this..., at least (if not even the `unicode` variable should work in a different way):

```diff
diff --git a/lib/plurimath/math/symbol.rb b/lib/plurimath/math/symbol.rb
index 70f8b37..7f81670 100644
--- a/lib/plurimath/math/symbol.rb
+++ b/lib/plurimath/math/symbol.rb
@@ -32,7 +32,7 @@ module Plurimath
         unicodes = Mathml::Constants::UNICODE_SYMBOLS
         unicode = unicodes.invert[value]
         if operator?(unicode) || unicode || explicit_checks(unicode)
-          mo_value = (unicodes[value] || unicode || value).to_s
+          mo_value = (unicodes[value.to_sym] || unicode || value).to_s
           return Utility.ox_element("mo") << mo_value
         end
 ```

...but then the tests are also incorrect. Based on the git history, you also seem to be an author of this fragment. Can you perhaps take a second look at this logic?

File: lib/plurimath/xml_engine/oga.rb (hunk 4 above)
Comment (hmdne) [reply lib/plurimath/xml_engine/oga.rb:206]: There's a check for `&& attr` on line 200 and 202, so unfortunately we can't use `case`.

**Comment by hmdne** [comment]
I have moved the last comment into a separate issue. Let's try to get this part merged. I will follow-up in a next comment.

**Comment by hmdne** [comment]
@ronaldtse @suleman-uzair 

I have made ahead-of-time Plurimath fully work with either Ox or Oga. To run tests under Ox, you can run `rake` just like before. To run tests under Oga, you can run `PLURIMATH_OGA=1 rake`.

Additionally, as requested, I have written a test case (xml_engine_spec.rb) that regardless of a setting mentioned above, tests the XML engine boilerplate for both engines.

There are only a couple of lines that are not covered now. One is an XPath edge case that I couldn't trigger consistently for both engines and isn't used by Plurimath currently, but from a perspective of having a full implementation makes sense for me. The other ones are guarded by a `RUBY_ENGINE == 'opal'` macro, therefore they aren't ran under MRI.

**Review by HassanAkbar — APPROVED** [review]
Looks good to me 👍

**Review by ronaldtse — APPROVED** [review]
Thank you @hmdne !
